packages/components/src/internal/components/lineage/LineageSettings.tsx

```diff
@@ -0,0 +1,216 @@
+import React, { ChangeEvent, FC, memo, useCallback, useMemo, useRef, useState } from 'react';
+
+import { SelectInput, SelectInputOption } from '../forms/input/SelectInput';
+
+import { LabelHelpTip } from '../base/LabelHelpTip';
+
+import { LINEAGE_GROUPING_GENERATIONS, LineageFilter, LineageOptions } from './types';
+import { DEFAULT_GROUPING_OPTIONS } from './constants';
+
+interface LineageSettingsOptions extends LineageOptions {
+    originalFilters?: LineageFilter[];
+}
+
+interface Props extends LineageOptions {
+    onSettingsChange: (options: LineageOptions) => void;
+    options?: LineageOptions;
+}
+
+export const LineageSettings: FC<Props> = memo(props => {
+    const { onSettingsChange } = props;
+    const [options, setOptions] = useState<LineageSettingsOptions>(() => {
+        if (props.options) return props.options;
+        return { filters: props.filters, grouping: props.grouping, originalFilters: props.filters };
+    });
+    const generations = useMemo<SelectInputOption[]>(() => {
+        const generations_ = [];
+        // eslint-disable-next-line guard-for-in
+        for (const value in LINEAGE_GROUPING_GENERATIONS) {
+            generations_.push({ label: value, value: LINEAGE_GROUPING_GENERATIONS[value] });
+        }
+        return generations_;
+    }, []);
+    const changeRef = useRef(undefined);
+
+    const applyOptions = useCallback(
+        (cb: (options: LineageSettingsOptions) => LineageSettingsOptions) => {
+            const options_ = cb(options);
+            setOptions(options_);
+
+            if (changeRef.current !== undefined) {
+                clearTimeout(changeRef.current);
+                changeRef.current = undefined;
+            }
+
+            changeRef.current = setTimeout(() => {
+                onSettingsChange(options_);
+            }, 500);
+        },
+        [onSettingsChange, options]
+    );
+
+    const onFilterChange = useCallback(
+        (evt: ChangeEvent<HTMLInputElement>) => {
+            const { checked, name } = evt.target;
+            applyOptions(options_ => {
+                const { originalFilters } = options_;
+                let { filters } = options_;
+
+                if (checked) {
+                    const filter = originalFilters.find(f => f.field === name);
+                    if (filter) {
+                        filters.push(filter);
+                    }
+                } else {
+                    filters = filters.filter(f => f.field !== name);
+                }
+
+                return { ...options_, filters };
```

Comment:
are you adding jest tests for this PR? if so, would it make sense to factor this out to be able to write unit tests for the checked / unchecked cases?

Reply:
Added a unit test.